Functionalize and clean up the Find_subterm implementation

ppedrot opened the pull request.

ppedrot commented:
@coqbot bench

SkySkimmer commented on pretyping/find_subterm.ml:
```diff
@@ -58,55 +58,87 @@ let map_named_declaration_with_hyploc f hyploc acc decl =
 type 'a testing_function = {
-  match_fun : 'a -> EConstr.constr -> 'a;
-  merge_fun : 'a -> 'a -> 'a;
+  match_fun : 'a -> EConstr.constr -> ('a, unit) Result.t;
```
Why not 'a option?

ppedrot replied:
Some intermediate state required a payload. Also, the intent is clearer to the user.

SkySkimmer replied:
The intermediate state doesn't matter.

>the intent is clearer to the user.

No it's not, we use option everywhere else for this sort of thing.

ppedrot replied:
Actually, the overreliance on option is a bit counterproductive at times. We use it for historical reasons, because the result types has only been around for a few OCaml versions (since 4.08 it seems). Option types should be used when the function has to return a dummy value, while result should be used to signal error instead. The fact that the error is dummy is not really relevant I think.

But if you insist I'll turn this into an option type...

ppedrot replied:
Actually I did find some dead code issues thanks to the use of result rather than option, I think it'd be better to keep it that way in the end.

Janno commented on pretyping/find_subterm.mli:
```diff
@@ -28,8 +26,8 @@ exception SubtermUnificationError of subterm_unification_error
     with None. *)
```
I think this comment is now outdated, isn't it?

ppedrot commented:
I've removed the last commit separating the checking part from the replacement part, it was making the code worse for no clear speedup. I've also removed some more dead code thanks to typing. I think this is ready.

SkySkimmer commented:
@coqbot merge now
